Testing

Benj1bear opened the pull request:
as per comments.py:

1. finish testing + other housekeeping
2. implement ternary handling
3. implement await handling (the only difference at a high level with it is that users will have an ```ag_await``` attribute they can access that tells them what is currently being awaited)

I'm essentially just going to finish testing, cleaning the code up, and adding/fixing the documentation. Ternary and await handling is kind of nieche or doesn't really add much of a difference for most users so I'll do that later.

Benj1bear committed:
Update comments.py

Fixed yieldfrom, recursion, closures, and global variables usage.

fixed recursions done in a closure so that the closure variable of the function name is correct.

fixed call so that it's calling to return objects not initializing a single object; preserving i.e. decorators and unintialized generators.

Benj1bear committed:
Update custom_generator.py

decided to make a BaseGenerator and use for the child classes Generator and AsyncGenerator since it seems in concept possible to do so but will need testing

corrected is_loop and is_definition for async for and async def respectively

Fixed aclose() method on AsyncGenerator

Benj1bear committed:
Update test_utils.py

added async tests
implemented ternary expressions
improved expr_getsource matching
fixed control_flow_adjust for multiple except catches
fixed generator expression handling in Generator.__init__
Cleaned up some of the code
fixed initialized generator lineno

Changed offset based tracking to instruction based for generator expressions. This now also means the minimum version supported is lower when excluding consideration for compound statements.

PabloRuizCuevas commented:
Cool , I will separate my implementation in another folder, and check a bit.  if you don´t mind could you squash commits and force push? otherwise will be hard to read the PR

Benj1bear commented:
I think I'll finish up with what I can to a reasonable extent on testing lambda expressions, go over clean source lines enough for most users, and then clean everything up this coming week.

It seems to work fine for most use cases but the value yields are creating some still probably fixable but nieche use cases that are going to require a lot more development time than I previously thought. I might create a separate pull request for those if needed.

Benj1bear committed:
Update test_utils.py

Decided to make a custom wrapper for dynamic bindings since fishhook only supports static bindings from what I've seen in their testing/ReadMe
Added support for lambda expressions
Added a function to determine if a builtin iterator type or our wrapper type is still running.
Fixed tracker issues

Benj1bear committed:
Update comments.py

Finished implementing ternary expressions
Fixed/cleaned some of the code up
Implemented multiple except catches
Fixed stack_adjuster for while loops
Fixed string_collector for f-strings
Fixed append_line for try-except-finally catches

Benj1bear commented:
I'm going to be away for a few weeks and may or may not work on this during then but not sure. All that's left to do is mostly testing/fixing ```unpack```, ```block_adjust```, and ```clean_source_lines``` in source_processing.py. I want to see this finished though so I'll be coming back to do some more commits until it's done.

Benj1bear commented:
I found something interesting:

I've just discovered that you can instantiate a frame and thus all that should be required is to work out how to cast its type into a generator type. It might not be possible but if it is possible it would save all the hacky functions that we've needed to implement so far:
https://stackoverflow.com/questions/79577323/copy-a-generator-via-ctypes

I'll still focus on finishing this project but if the ctypes option works out then that'll suffice.

Benj1bear commented:
Not sure when I'll be back. I'm going through a lot at the moment but will be coming back to finish when I've got time. I also posted my thoughts on your post on the python.org discussion so it'll be interesting what feedback they have for me.

PabloRuizCuevas commented:
@Benj1bear Just read this, i had also quite a bunch of things going on, but I may be able to push a bit more now. Also I was looking at your post in the python forum, as you may see attracting the attention there is complicated.

But I think the correct approach is packaging whatever we have now in pypy and publish it in a new post (rather than using my old one that is kind of burned for the effects of that forum).  In the post shoring working examples is extremely important, once we do this I think we will attract some attention.

What do you think? I think pretending to have the perfect solution is out of the scope, but you got already something very cool here, so we can start from that without more modifications and then let's see if we improve it in the future. Are you familiar with publishing packages? i can do it myself too, as you wish.

Once is published, i think you can make a post in the forum (a new one) which i will support, (i can also do the new post, but i think at this point you deserve way more credit than i do). And finally we can go to a couple of SO old questions and post this new solution there.

if this works in 70% of the normal cases is already a major major contribution to the python community, there is so many things i wanted to pickle a generator i can´t event count them. there are so many questions of people confused because they don´t know why their multiprocessing code doesn´t work because this issue... and so many others issues that are solved by this!

Benj1bear commented:
Agreed, well put.

I'm unfamiliar with packaging to pypi so I wouldn't mind if you did this that'd be great.

I'm not too familiar with python.org type forum expectations so I think if you have in mind what approach will do well you should make the post and I can help where needed. I'm happy with knowing the work we've done and built upon will increase the capacity and possibilities of others work which is exciting.

PabloRuizCuevas commented:
@Benj1bear Ok I will take care of the packaging and advertising it in the forum. I will publish it on monday. If you have a name for the package is the time to tell me, some options are:  copygen, deepgen, genclone, genstate, picklegen, do you like any of them?

I don´t really know for certain what will do well in the forum haha as you see they are hard, but will try my best for making it as appealing as possible, in anyway putting something inside python is a long distance run, i expect it will only ever be built in if it get first some real usage from the community, and for that people need to know that now it is possible.


 If I can ask you a last thing it would be to create a couple of examples to put in the readme, but i may find time to do it myself as well, so no pressure.

Benj1bear commented:
Cool. I liked the name you gave the directory e.g. gcopy but otherwise out of the examples copygen is fine. I can add some examples to the readme, I'll likely be back later in a few days and can do that then.

PabloRuizCuevas commented:
@Benj1bear I Merged the PR, I'm doing some small changes we need , specially fixing imports, as import * is almost a no go nowadays and made everything a bit confusing in my opinion, Specially when things have the same name like Generator (our own or the one from typing) is hard to know where it is coming from.

Anyway I'm not yet finished with the fixed but in case you want to check, I advise you to run "pytest ." to run the tests and see how is it doing.

later today or tomorrow will try to fix the missing tests and finalize the packaging.|

Also I moved my code to etc, and placed a license file and part of your readme in the root

PabloRuizCuevas commented:
@Benj1bear I fixed most tests (i think 2-3 of them were already not working) and maybe i broke a couple of them. but some questions arised.

tests_picklers -> not sure why it fails, maybe you can take a look.

test_generator__iter__ -> fails but worst than that has this line `assert [i for i in gen] == [1, 2] `  why is `gen and not gen()` is not called like a normal generator? could you clarify me this point pls?


test_gen_expr, test_lambda_expr -> maybe were already failing? not sure, but also not sure how fix them.

test_track_iter_inside_exec, test_track_iter_inside_Generator -> sometimes fail sometimes not, i feel they depend on the globals and when running in isolation fail? maybe

test_get_globals, test_is_running -> not sure if fail for similar reasons as the previous

Besides that the refactor for packaging is  almost done we just need to clarify these last things. 98/105 test passing

PabloRuizCuevas commented:
Also I'm considering renaming  Generator, Is not great that overlaps the name of the built-in one, unless we want to directly monkey patch it and substitute it, it kind of pollutes the namespace as it avoids to use normal generators without renaming it. I will give it a think...

PabloRuizCuevas commented:
Also: 

```python
from gcopy.custom_generator import Generator
@Generator
def gen_test():
    b = 1
    while True:
        yield b
        print(b)
        b +=1

t= gen_test()
next(t)
next(t)
next(t)
```

this is not working, which is quite basic, it just returns 1, 1, 1, not sure now if i broke it myself.

PabloRuizCuevas commented:
I created a jupyter called examples with this one, in case you have a second to take a look on these

Benj1bear commented:
I'll see what's going on

PabloRuizCuevas commented:
good thanks! and remember calling the test using "pytest ." some issues only appear in that way or sometimes "pytest <the specific test>"

Benj1bear commented:
I checked the merge e.g. git reset --hard 347c53f1e0280b4c62c6165c5af251c74d2cd87c
and ran the tests and it's passing.

The gen without brackets in test_generator__iter__  is because I'm wrapping a function with a class to make a generator instance and regardless of it being called it can still run ```__next__``` e.g. basically gen == gen() which is weird and I probably should've written a comment there ... am wondering if for some reason I left it like that, can't remember why I did that. So it's something I need to modify when doing function generators. 

Changing line 367 in gcopy/custom_generator.py can prevent this from occurring:
```python
if isinstance(FUNC, FunctionType):
    class CallError:
        def __next__(self):
            raise TypeError("Cannot iterate an uninitialized function generator")

    self._internals["state_generator"] = CallError()
else:
    self._internals["state_generator"] = self._init_states()
```

or init_states can be modified:
```python
    def _init_states(self, initialized: bool = True) -> GeneratorType:
        """Initializes the state generation as a generator"""
        if not initialized:
            raise TypeError("Cannot iterate an uninitialized function generator")
        self._internals["loops"] = get_loops(self._internals["lineno"], self._internals["jump_positions"])
        ## if no state then it must be EOF ##
        while self._internals["state"]:
            yield self._create_state()

```

but users can still manually override this but at that point it's on them since they want to change the internals.

Some tests will likely need to be rewritten for this change.

I'll run your changes now.

Benj1bear commented:
I ran tests with your changes (not using pytest, will try it next) and it still works. One thing I should mention is that I move all the .py files in the test folder up a directory and then run them there because I wasn't sure how to get the imports working relative or however they needed to be for that test directory which is likely an issue.

Also, I ran the example notebook as follows and it works (I still ran ```cd .. ``` though not shown):

![c1](https://github.com/user-attachments/assets/83950ac6-92a5-4d29-84c2-6d7f0178d9a2)

We probably need ```patch_iterators()``` on ```__init__``` of the module if we don't want users to want to think about it.

But the other example you've given I think is a legitimate issue. Maybe I haven't checked while loops (I can't remember) but it should be clear what's going on from looking at the states source code so I'll look into it.

![c2](https://github.com/user-attachments/assets/ec849935-627f-44dc-bdaa-b2b2829b43a7)

Benj1bear commented:
The issue is either with the line position given to it or slicing on adjustment. Will have a look to see which one and make sure it's not something else.
![image](https://github.com/user-attachments/assets/38da8067-fa6e-4129-ae12-cf7708ddeedd)
![image](https://github.com/user-attachments/assets/97ffbd38-83ce-4a24-a5ca-037657a55d86)
![image](https://github.com/user-attachments/assets/ea1a7504-d59d-4a83-ad68-d73283db90f6)

Benj1bear commented:
It's also not limited to while loops it happens on for loops as well. Will look into where it's coming from.

PabloRuizCuevas commented:
@Benj1bear  ah ok i forgot the patch_iterators(), but maybe we need just to do that under the hood, not elegant but better than putting it in user space. for the rest I will check it later with more time, if you made any change feel free to start another PR

Benj1bear commented:
yes you can do that in _frame_init at line 497 in gcopy/custom_generator.py. You'd need to initially assume that the user wants the builtin iterators patched. If they choose to monkey patch these themselves then their overrides will override our overrides:

```python
  ...
  
  # keep separate from user space
  patched_iterators = dict()
  patch_iterators(patched_iterators)
  
  class BaseGenerator(Pickler):
      ...
      def _frame_init(...) -> ...:
          ...
          # ensure patched_iterators are the initial assumption, user overrides should still take into effect this way
          exec(code_obj, patched_iterators | self._internals["frame"].f_globals, locals())
          ...
      ...
```

But for already running generators they will need to run ```patch_iterators()``` into their scope otherwise the implicit iterators are not tracked.

Benj1bear commented:
I noted that:
```python
## to ensure gcopy.custom_generator.Generator can be used in exec for sign ##
import gcopy
```
was moved down and changed to:
```python
## TODO, somehow used in Generator
import gcopy
#
```
There's an exec used in the ```gcopy.source_processing.sign``` function essentially to dynamically create a new function that requires ```gcopy``` be available since it makes a call to ```gcopy.custom_generator.Generator```. The ```sign``` function is used to make the ```Generator__call__``` function appear as the actual function call.

Also, I think I might have made some changes since I last tested ```test_track_iter_inside_exec``` because it doesn't seem to test anything it supposed to e.g. to do with ```track_iter``` inside exec but this is covered in ```test_track_iter_inside_Generator``` anyway so if really needed we can remove ```test_track_iter_inside_exec```.

Benj1bear commented:
I've applied the changes.

I'm going to:
 - Fix the found slicing issue
 - Add the functions fix
 - rewrite some tests
 
Then afterwards I'll write some examples as well

PabloRuizCuevas commented:
This thing of the exec, ideally the import should be inside the function where the exec lives and maybe the functions passed to the globals or locals of the exec, something that is very scoped so we don´t have import traps that can´t be deleted and you don´t know why. I couldn´t find the exec you mentioned, but I imagined it would be there.

but yes feel free  to open a new PR with these changes. Thanks :D

Benj1bear commented:
ah I found out what it was, basically it was trying to get the type annotations when it was signing the function and one of the return annotations was gcopy.custom_generator.Generator but instead I can just manually set them and that should make it scoped. Will add this.

Benj1bear commented:
I'm thinking about that implicit loop stuff. It might be possible to inspect the c level variables on the interpreters stack that's found in all frames but is not exposed at run time so would have to cast the frame into a custom ctypes.Structure with the correct fields to reveal them:

https://github.com/python/cpython/blob/3358f9abf4d05174486d6aadf464b5849b25be7c/Include/cpython/frameobject.h#L28C1-L50C3

Though it seems in newer versions the memory layout of frames have changed so will need to be careful.

Will see if this is possible later.
